var/spack/repos/builtin/packages/h5bench/package.py

```diff
@@ -58,3 +62,40 @@ def cmake_args(self):
         ]
 
         return args
+
+    @run_after("install")
+    def setup_build_tests(self):
+        launcher = self.mpi_launcher()
+
+        exe = os.path.basename(launcher.path)
+        filename = "samples/sync-write-1d-contig-contig.json"
+        filter_file(f"{exe}", f"{launcher}", filename)
+        if self.spec["slurm"]:
+            filter_file(r"mpirun", "srun", filename)
+            filter_file(r"--allow-run-as-root -n 2", "-n 1", filename)
+        else:
+            filter_file(r"--allow-run-as-root -n 2", "-n 1 --timeout 240", filename)
+
+        """Copy the example source files after the package is installed to an
+        install test subdirectory for use during `spack test run`."""
+        cache_extra_test_sources(self, ["tests", "samples"])
+
+    def mpi_launcher(self):
+        searchpath = [self.spec["mpi"].prefix.bin]
+        try:
+            searchpath.insert(0, self.spec["slurm"].prefix.bin)
+        except KeyError:
+            print("Slurm not found, ignoring.")
+        commands = ["srun", "mpirun", "mpiexec"]
+        return which(*commands, path=searchpath) or which(*commands)
```

Comment:
The following implementations of `setup_build_test` and `mpi_launcher` work for me (note I also removed `slurm` as a dependency):

```suggestion
    @run_after("install")
    def setup_build_tests(self):
        launcher = self.mpi_launcher()
        filename = "samples/sync-write-1d-contig-contig.json"
        filter_file(f"mpirun", f"{launcher}", filename)
        filter_file(r"-n 2", "-n 1 --timeout 240", filename)

        """Copy the example source files after the package is installed to an
        install test subdirectory for use during `spack test run`."""
        cache_extra_test_sources(self, ["tests", "samples"])

    def mpi_launcher(self):
        commands = ["mpirun", "mpiexec"]
        return which(*commands, path=[self.spec["mpi"].prefix.bin]) or which(*commands)
```

Reply:
With these changes, I get the following test results after installing the software:

```
spack -v test run h5bench
==> Spack test sp7tpic6lqfruu3mn73b6hastmry5ieg
==> Testing package h5bench-1.4-mnhye77
==> [2023-08-03-15:39:29.027630] Installing SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-mnhye77ygp55ogjw54fh4muwvvaxm73y/.spack/test to SPACK_TEST_ROOT/sp7tpic6lqfruu3mn73b6hastmry5ieg/h5bench-1.4-mnhye77/cache/h5bench
==> [2023-08-03-15:39:29.615168] test: test_h5bench: Run h5bench synchronous write test.
==> [2023-08-03-15:39:29.619846] 'SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-mnhye77ygp55ogjw54fh4muwvvaxm73y/bin/h5bench' '--debug' '--abort' 'samples/sync-write-1d-contig-contig.json'
2023-08-03 15:39:30,007 h5bench - INFO - Starting h5bench Suite
2023-08-03 15:39:30,019 h5bench - INFO - Lustre support not detected
2023-08-03 15:39:30,020 h5bench - DEBUG - LD_LIBRARY_PATH: 
2023-08-03 15:39:30,024 h5bench - DEBUG - DYLD_LIBRARY_PATH: 
2023-08-03 15:39:30,027 h5bench - DEBUG - LD_PRELOAD: 
2023-08-03 15:39:30,029 h5bench - INFO - h5bench [write] - Starting
2023-08-03 15:39:30,033 h5bench - INFO - h5bench [write] - DIR: storage/a6e560b7/
2023-08-03 15:39:30,035 h5bench - INFO - Parallel setup: SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/openmpi-4.1.5-zq5vqifl2tassijpbyeekjj4onaqvxsc/bin/mpirun --allow-run-as-root -n 1 --timeout 240 --oversubscribe
2023-08-03 15:39:30,039 h5bench - INFO - SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/openmpi-4.1.5-zq5vqifl2tassijpbyeekjj4onaqvxsc/bin/mpirun --allow-run-as-root -n 1 --timeout 240 --oversubscribe SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-mnhye77ygp55ogjw54fh4muwvvaxm73y/bin//h5bench_write storage/a6e560b7/h5bench.cfg storage/test.h5
2023-08-03 15:39:42,075 h5bench - INFO - SUCCESS (all output files are located at storage/a6e560b7)
2023-08-03 15:39:42,076 h5bench - INFO - Runtime: 12.0404091 seconds (elapsed time, includes allocation wait time)
2023-08-03 15:39:42,077 h5bench - INFO - h5bench [write] - Complete
2023-08-03 15:39:42,078 h5bench - INFO - Finishing h5bench Suite
PASSED: H5bench::test_h5bench
==> [2023-08-03-15:39:42.091990] test: test_help: Run h5bench help.
==> [2023-08-03-15:39:42.094824] 'SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-mnhye77ygp55ogjw54fh4muwvvaxm73y/bin/h5bench' '-h'
usage: h5bench [-h] [-a] [-d] [-v] [-p PREFIX] [-f FILTER] [-V] setup

H5bench: a Parallel I/O Benchmark Suite for HDF5:

positional arguments:
  setup                 JSON file with the benchmarks to run

optional arguments:
  -h, --help            show this help message and exit
  -a, --abort-on-failure
                        Stop h5bench if a benchmark failed
  -d, --debug           Enable debug mode
  -v, --validate-mode   Validated if the requested mode (async/sync) was run
  -p PREFIX, --prefix PREFIX
                        Prefix where all h5bench binaries were installed
  -f FILTER, --filter FILTER
                        Execute only filtered benchmarks
  -V, --version         show program's version number and exit
PASSED: H5bench::test_help
==> [2023-08-03-15:39:42.352442] Completed testing
==> [2023-08-03-15:39:42.352606] 
========================= SUMMARY: h5bench-1.4-mnhye77 =========================
H5bench::test_h5bench .. PASSED
H5bench::test_help .. PASSED
============================= 2 passed of 2 parts ==============================
============================== 1 passed of 1 spec ==============================
```

Reply:
You can see from the test output when I was trying to use the system-installed slurm:

```
$ spack -v test run h5bench
==> Spack test k6fxq2xxxnyivki65gutedzpj4d2v6yw
==> Testing package h5bench-1.4-klu4xhu
==> [2023-08-03-15:23:44.458582] Installing SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-klu4xhulpw4zi4yxpurkkckuw4z2g7ok/.spack/test to SPACK_TEST_ROOT/k6fxq2xxxnyivki65gutedzpj4d2v6yw/h5bench-1.4-klu4xhu/cache/h5bench
==> [2023-08-03-15:23:45.257058] test: test_h5bench: Run h5bench synchronous write test.
==> [2023-08-03-15:23:45.261211] 'SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-klu4xhulpw4zi4yxpurkkckuw4z2g7ok/bin/h5bench' '--debug' '--abort' 'samples/sync-write-1d-contig-contig.json'
2023-08-03 15:23:45,608 h5bench - INFO - Starting h5bench Suite
2023-08-03 15:23:45,620 h5bench - INFO - Lustre support not detected
2023-08-03 15:23:45,620 h5bench - DEBUG - LD_LIBRARY_PATH: 
2023-08-03 15:23:45,621 h5bench - DEBUG - DYLD_LIBRARY_PATH: 
2023-08-03 15:23:45,622 h5bench - DEBUG - LD_PRELOAD: 
2023-08-03 15:23:45,623 h5bench - INFO - h5bench [write] - Starting
2023-08-03 15:23:45,624 h5bench - INFO - h5bench [write] - DIR: storage/53129e3a/
2023-08-03 15:23:45,625 h5bench - INFO - Parallel setup: /usr/bin/srun -n 1 --oversubscribe
2023-08-03 15:23:45,629 h5bench - INFO - /usr/bin/srun -n 1 --oversubscribe SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-klu4xhulpw4zi4yxpurkkckuw4z2g7ok/bin//h5bench_write storage/53129e3a/h5bench.cfg storage/test.h5
2023-08-03 15:24:04,478 h5bench - ERROR - Return: 1 (check storage/53129e3a/stderr for detailed log)
2023-08-03 15:24:04,479 h5bench - CRITICAL - h5bench execution aborted upon first error
FAILED: H5bench::test_h5bench: Command exited with status 70:
    'SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-klu4xhulpw4zi4yxpurkkckuw4z2g7ok/bin/h5bench' '--debug' '--abort' 'samples/sync-write-1d-contig-contig.json'
...
==> [2023-08-03-15:24:04.806736] Completed testing
==> [2023-08-03-15:24:04.806900] 
========================= SUMMARY: h5bench-1.4-klu4xhu =========================
H5bench::test_h5bench .. FAILED
H5bench::test_help .. PASSED
======================== 1 failed, 1 passed of 2 parts =========================
```

Reply:
And the failure when I tried to use the full path of the `slurm` dependency installed with the package:

```
$ spack -v test run h5bench
==> Spack test bdjmazff55pypfpyq7hnvxpucnxtrplu
==> Testing package h5bench-1.4-jryyrgt
==> [2023-08-03-15:05:54.783461] Installing SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-jryyrgt5khbpdl2mpxfoo6pkka2g6xgx/.spack/test to SPACK_TEST_ROOT/bdjmazff55pypfpyq7hnvxpucnxtrplu/h5bench-1.4-jryyrgt/cache/h5bench
==> [2023-08-03-15:05:55.263188] test: test_h5bench: Run h5bench synchronous write test.
==> [2023-08-03-15:05:55.264889] 'SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-jryyrgt5khbpdl2mpxfoo6pkka2g6xgx/bin/h5bench' '--debug' '--abort' 'samples/sync-write-1d-contig-contig.json'
2023-08-03 15:05:55,488 h5bench - INFO - Starting h5bench Suite
2023-08-03 15:05:55,504 h5bench - INFO - Lustre support not detected
2023-08-03 15:05:55,505 h5bench - DEBUG - LD_LIBRARY_PATH: 
2023-08-03 15:05:55,506 h5bench - DEBUG - DYLD_LIBRARY_PATH: 
2023-08-03 15:05:55,507 h5bench - DEBUG - LD_PRELOAD: 
2023-08-03 15:05:55,508 h5bench - INFO - h5bench [write] - Starting
2023-08-03 15:05:55,509 h5bench - INFO - h5bench [write] - DIR: storage/ca246819/
2023-08-03 15:05:55,510 h5bench - INFO - Parallel setup: SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/slurm-23-02-2-1-oakj6a4lceqlyxjwltqzuakuz635ex4b/bin/srun -n 1 --timeout 240
2023-08-03 15:05:55,514 h5bench - INFO - SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/slurm-23-02-2-1-oakj6a4lceqlyxjwltqzuakuz635ex4b/bin/srun -n 1 --timeout 240 SPACK_ROOT/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/h5bench-1.4-jryyrgt5khbpdl2mpxfoo6pkka2g6xgx/bin//h5bench_write storage/ca246819/h5bench.cfg storage/test.h5
2023-08-03 15:05:55,535 h5bench - ERROR - Return: 1 (check storage/ca246819/stderr for detailed log)
2023-08-03 15:05:55,535 h5bench - CRITICAL - h5bench execution aborted upon first error
FAILED: H5bench::test_h5bench: Command exited with status 70:
...
PASSED: H5bench::test_help
==> [2023-08-03-15:05:55.795010] Completed testing
==> [2023-08-03-15:05:55.795290] 
========================= SUMMARY: h5bench-1.4-jryyrgt =========================
H5bench::test_h5bench .. FAILED
H5bench::test_help .. PASSED
======================== 1 failed, 1 passed of 2 parts =========================
```

Reply:
I've tested this locally and on Perlmutter, and it seems to have solved the issue. Thanks!